app/src/org/commcare/activities/HomeScreenBaseActivity.java

```diff
@@ -189,15 +194,86 @@ private void loadInstanceState(Bundle savedInstanceState) {
     }
 
     /**
-     * Set state that signifies activity was launch from external app.
+     * Set state that signifies activity was launch from external app
      */
     private void processFromExternalLaunch(Bundle savedInstanceState) {
         if (savedInstanceState == null && getIntent().hasExtra(DispatchActivity.WAS_EXTERNAL)) {
             wasExternal = true;
+            processSessionEndpoint();
             sessionNavigator.startNextSessionStep();
         }
     }
 
+    private void processSessionEndpoint() {
+        if (getIntent().hasExtra(SESSION_ENDPOINT_ID)) {
+            Endpoint endpoint = validateIntentForSessionEndpoint(getIntent());
+            if (endpoint != null) {
+                Vector<String> endpointArguments = endpoint.getArguments();
+                EvaluationContext evaluationContext = CommCareApplication.instance().getCurrentSessionWrapper().getEvaluationContext();
+
+                Bundle intentArgumentsAsBundle = getIntent().getBundleExtra(SESSION_ENDPOINT_ARGUMENTS_BUNDLE);
+                ArrayList<String> intentArgumentsAsList = getIntent().getStringArrayListExtra(SESSION_ENDPOINT_ARGUMENTS_LIST);
+
+                for (int i = 0; i < endpointArguments.size(); i++) {
+                    String argumentName = endpointArguments.elementAt(i);
+                    if (intentArgumentsAsBundle != null && !intentArgumentsAsBundle.containsKey(argumentName)) {
+                        String noArgumentWithNameError = org.commcare.utils.StringUtils.getStringRobust(
+                                this,
+                                R.string.session_endpoint_no_argument_with_name,
+                                new String[]{argumentName, endpoint.getId()});
+                        UserfacingErrorHandling.createErrorDialog(this, noArgumentWithNameError, true);
+                        return;
+                    }
+
+                    // Bundle Arguments take precdence over list arguments
+                    String argValue = intentArgumentsAsBundle == null ? intentArgumentsAsList.get(i)
+                            : intentArgumentsAsBundle.getString(argumentName);
+
+                    evaluationContext.setVariable(argumentName, argValue);
+                }
+
+                CommCareApplication.instance().getCurrentSessionWrapper().reset();
+                CommCareApplication.instance().getCurrentSessionWrapper()
+                        .executeStackActions(endpoint.getStackOperations(), evaluationContext);
+            }
+        }
+    }
+
+    private Endpoint validateIntentForSessionEndpoint(Intent intent) {
```

Comment:
do we need the check block on line 219 if this validation code runs over things first?